pkg/p2p/libp2p/libp2p.go

```diff
@@ -358,12 +359,20 @@ func (s *Service) handleIncoming(stream network.Stream) {
 	}
 
 	if i.FullNode {
-		err = s.addressbook.Put(i.BzzAddress.Overlay, *i.BzzAddress)
-		if err != nil {
-			s.logger.Debugf("stream handler: addressbook put error %s: %v", peerID, err)
-			s.logger.Errorf("stream handler: unable to persist peer %v", peerID)
-			_ = s.Disconnect(i.BzzAddress.Overlay, "unable to persist peer in addressbook")
-			return
+
+		ctx, cancel := context.WithTimeout(context.Background(), pingTimeout)
+		_, err := s.Ping(ctx, i.BzzAddress.Underlay)
+		cancel()
+
+		// only insert to addressbook if peer is reachable
+		if err == nil {
+			err = s.addressbook.Put(i.BzzAddress.Overlay, *i.BzzAddress)
```

Comment:
The issue here is although the peer's underlay address is not reachable, we have an active connection with it. If we dont put the address in the addressbook, we have an overlay address in the known peers list and not in the addressbook. Is this okay? Can this cause issues elsewhere?

Reply:
In this case that this peer disconnects, because this peer is in the knownpeers list, kademlia could possibly attempt to dial for a connection again in the future, so good catch. 

I will fix this.

Reply:
the other thing is that we won't be announcing this peer since the addressbook entry could not be found. so this peer will be picked up and would return an error when we try to announce. this could result in discovery issues, since we gossip less peers than expected

Reply:
The current hive logic skips the peer if the addressbook entry is not found and continues with other peers. 
In general, afaik unreachable peers should not be announced.

Reply:
hive will skip but kademlia gives hive a peer which is not consistent with the address book state. i.e. kademlia assumes that the subset of peers given for discovery are all in the address book but this is not the case.

Reply:
what i'm saying here is that you're adding more dependencies within kademlia when you do the announcements, since you can't just pick `n` nodes to gossip randomly, but you must ensure they are all in the address book prior to communicating them to hive

Reply:
Ideally, all entries in the addressbook should be reachable. This is why we do pinging now before we insert peers into the addressbook.
Since not all connected peers in kademlia may be reachable,  kademlia cannot make that assumption.

Reply:
this is not about reachability, please have a look at `kademlia.go#Announce`:
```go
// Announce a newly connected peer to our connected peers, but also
// notify the peer about our already connected peers
func (k *Kad) Announce(ctx context.Context, peer swarm.Address, fullnode bool) error {
	var addrs []swarm.Address

	for bin := uint8(0); bin < swarm.MaxBins; bin++ {

		connectedPeers, err := randomSubset(k.connectedPeers.BinPeers(bin), broadcastBinSize)
		if err != nil {
			return err
		}

		for _, connectedPeer := range connectedPeers {
			if connectedPeer.Equal(peer) {
				continue
			}

			addrs = append(addrs, connectedPeer)

			if !fullnode {
				// we continue here so we dont gossip
				// about lightnodes to others.
				continue
			}
			go func(connectedPeer swarm.Address) {
				if err := k.discovery.BroadcastPeers(ctx, connectedPeer, peer); err != nil {
					k.logger.Debugf("could not gossip peer %s to peer %s: %v", peer, connectedPeer, err)
				}
			}(connectedPeer)
		}
	}
```
here, `randomSubset` picks a random few connected peers, however with your change it is no longer guaranteed that any of these peers are in the address book, resulting in peers not being gossiped to other peers